### ISSUE TITLE
Add workaround for buggy ssh password prompt behavior in macOS 10.15.4

### ISSUE THREAD

**Issue by ralfdaniel** [bcpierce00/unison#328]
Hi,
I am syncing my MacBook and iMac data folders with Unison 2.51.2 for several years now and it worked great.
But after updating macOS Catalina from version 10.15.3 to 10.15.4 released on 23rd March 2020 Unison is not able to connect to the second computer. It is just showing "Connecting..." and nothing more happens.
I already checked the SSH connection between both computer via Terminal and it is working in both directions. Also there was no change in the .prf-file.
Does some has any fix or idea how to solve this?

**Comment by samuel** [comment]
Have the same issue after upgrading to 10.15.4. Seems to be `2020-03-30 09:13:37.421 Unison[10125:80781] Unrecognized message from ssh: ` 

Running the unison cli installed through homebrew works. So it seems to be something to do with the release build. I'm happy to help track this down. Currently trying to build the app to add more debugging.

Full output of `-debug all` doesn't seem to show anything useful.

```
Preferences:
ui = graphic
host = 
server = false
prefsdocs = false
doc = 
version = false
silent = false
dumbtty = false
testserver = false
showprev = false
selftest = false
confirmmerge = false
retry = 0
repeat = 
contactquietly = false
key = 
label = 
expert = false
height = 15
auto = false
maxthreads = 0
maxsizethreshold = -1
prefer = 
force = 
sortnewfirst = false
sortbysize = false
keeptempfilesaftermerge = false
diff = diff -u CURRENT2 CURRENT1
copyonconflict = false
backupdir = 
maxbackups = 2
backups = false
backupsuffix = 
backupprefix = .bak.$VERSION.
backuploc = central
copymax = 1
copyquoterem = default
copythreshold = -1
copyprogrest = rsync --partial --append-verify --compress
copyprog = rsync --partial --inplace --compress
rsync = true
fastcheck = default
ignorelocks = false
dumparchives = false
showarchive = false
rootsName = 
ignorearchives = false
fastercheckUNSAFE = false
fat = false
allHostsAreRunningWindows = false
someHostIsRunningWindows = false
confirmbigdel = true
batch = false
root = ssh://samuel@192.168.84.26//home/samuel/unison
root = /Users/samuel/unison
killserver = false
halfduplex = false
stream = true
addversionno = false
servercmd = 
sshargs = 
rshargs = 
rshcmd = rsh
sshcmd = ssh
xferbycopying = true
sshversion = 
clientHostName = mycomp.local
ignoreinodenumbers = false
links-aux = true
links = default
times = false
group = false
owner = false
numericids = false
dontchmod = false
perms = 1023
watch = true
rsrc-aux = false
rsrc = default
maxerrors = 1
unicodeCS = false
unicodeEnc = false
unicode = default
someHostIsInsensitive = false
ignorecase = default
timers = false
terse = false
logfile = /Users/samuel/unison.log
log = true
debugtimes = false
debug = all
addprefsto = 
[remote] Shell connection: ssh (ssh, -l, samuel, 192.168.84.26, -e, none, unison, -server)
2020-03-30 09:15:43.804 Unison[10150:81885] Unrecognized message from ssh: 
```

**Comment by samuel** [comment]
Using the text ui works but the mac ui doesn't. Seems to be maybe something with pipes and the ssh command but not totally sure. At least this is a work around for now.

```
$ /Applications/Unison.app/Contents/MacOS/Unison -ui text Home                                                                                                2020-03-31 17:59:50.520 Unison[46737:626400] Calling nonGuiStartup
Unison 2.51.2 (ocaml 4.04.0): Contacting server...
password: 

$ /Applications/Unison.app/Contents/MacOS/Unison Home\ Go                                                                                                        2020-03-31 17:59:58.926 Unison[46750:626495] Unrecognized message from ssh: 
```

**Comment by ralfdaniel** [comment]
Great, thank you Samuel! This also did the job to sync my macs. So it must be something with the mac ui if command line entries are working.

**Comment by bcpierce00** [comment]
Using "-debug verbose” instead of “-debug all” will print the exact bytes that the client is getting from the ssh process.  This may help understand whether the problem is something to do with the communication between the client and the ssh process or something else.

   - Benjamin

**Comment by h2k82u** [comment]
Same problem here, Unison not working since the update to 10.15.4.
ssh seems to have a bug / different behaviour since the update:
https://tyler.io/so-uh-i-think-catalina-10154-broke-ssh/
BTW: I use port 22 for the ssh server

**Comment by h2k82u** [comment]
Still not working after 'macOS Catalina 10.15.4 Supplemental Update' on both machines
Unison gui is freezing and showing 'Connecting'
I would be glad to help testing and tracking that error

**Comment by samuel** [comment]
> Using "-debug verbose” instead of “-debug all” will print the exact bytes that the client is getting from the ssh process. This may help understand whether the problem is something to do with the communication between the client and the ssh process or something else. - Benjamin


Using `-debug verbose` is showing the same output for me:

```
logfile = /Users/samuel/unison.log
log = true
debugtimes = false
debug = verbose
addprefsto = 
[remote] Shell connection: ssh (ssh, -l, samuel, 192.168.84.26, -e, none, unison, -server)
2020-04-09 11:54:55.755 Unison[2220:22793] Unrecognized message from ssh: 
```

**Comment by h2k82u** [comment]
log = true
debugtimes = false
debug = verbose
addprefsto = 
[remote] Shell connection: ssh (ssh, -l, bs, 192.168.178.7, -e, none, /usr/local/bin/unison, -server)
2020-04-10 08:44:46.053 Unison[5178:185556] Unrecognized message from ssh:

**Comment by bcpierce00** [comment]
Looks like this message is coming from line 489 of uimac/MyController.m and the message from ssh that it can’t interpret is the empty string (or a string containing just blanks).  

If somebody wants to do a little objective C programming, one could try to protect all the calls to raisePasswordWindow with a little loop that throws away empty prompts from ssh.

**Comment by etziperman** [comment]
same problem for me...  both using pre-compiled binary and macports version. and the text interface works for me too.

**Comment by g8u** [comment]
Open a shell and use ssh-add
This solves the issue for me.

**Comment by kevinashworth** [comment]
I'm on Catalina 10.15.4. I typed `ssh-add` at the command line, it added my identity. Then I tried to run Unison, but to no avail -- still just shows "Connecting...". But I can continue to use the command-line text version of Unison.
Hmm. Wonder why it worked for g8u but not me?

**Comment by g8u** [comment]
> I'm on Catalina 10.15.4. I typed `ssh-add` at the command line, it added my identity. Then I tried to run Unison, but to no avail -- still just shows "Connecting...". But I can continue to use the command-line text version of Unison.
> Hmm. Wonder why it worked for g8u but not me?

I am clueless. Just can mention I sync between Catalina 10.15.4 and Ubuntu 20.04.

**Comment by h2k82u** [comment]
@g8u:
can you try syncing between two Catalina machines?
As far as I understand the problem is what the second machine (in your case Ubuntu?) reports back over ssh

**Comment by g8u** [comment]
> @g8u:
> can you try syncing between two Catalina machines?
> As far as I understand the problem is what the second machine (in your case Ubuntu?) reports back over ssh

(un)fortunately I haven't got a second Mac. As far as I understand  Unison shows an error if it receives an unexpected answer (as I just experienced painfully until getting Unison working again in my environment)

**Comment by etz1** [comment]
I am synching with a previous version OS Mac and this solution unfortunately did not help. The identity was added, but the unison window is still stuck saying "connecting" and does not prompt for a password. the text version works well, fortunately.

**Comment by g8u** [comment]
> I'm on Catalina 10.15.4. I typed `ssh-add` at the command line, it added my identity. Then I tried to run Unison, but to no avail -- still just shows "Connecting...". But I can continue to use the command-line text version of Unison.
> Hmm. Wonder why it worked for g8u but not me?

To find out, whether this is an issue with RSA passphrase handling when
unison connects to the host or a problem occuring later on you might
want to test this by temporarily removing the passphrase from your RSA key:

$ ssh-keygen -p

https://stackoverflow.com/questions/112396/how-do-i-remove-the-passphrase-for-the-ssh-key-without-having-to-create-a-new-ke

**Comment by etz1** [comment]
SOLVED! for me...:

I ssh from a terminal to the machine with which I would like to sync using unison. I have a setup (.ssh/config) that allows me to then ssh from a different terminal window with no password to that remote machine, as long as the first ssh command is active. more on that below.

I start Unison, graphic version, it connects without requiring a password and works well. The remote hostname in the .unison file needs to be the same name that I use for the first ssh command.

as for the .ssh/config setup, here it is, probably only some of this is relevant:

```
Host my.desktop.company.com
	User myuserid
	HostName my.desktop.company.com
	ControlMaster auto
	ControlPath ~/.ssh/%r@%h:%p
	ServerAliveInterval 2400
   	ForwardAgent yes
    	ForwardX11 yes
    	ForwardX11Trusted yes
    	#Port 22
    	#Protocol 2
	XauthLocation /opt/X11/bin/xauth
    	GSSAPIDelegateCredentials no

```

**Comment by kevinashworth** [comment]
Thanks! This works for me, too. Following your instructions, I ssh in first, using the same hostname as in my .prf file, and then the Unison graphical version works as before, and after syncing I end my ssh session after closing Unison. 

The contents of my `.ssh/config` are just this:

```
Host *
   ControlMaster auto
   ControlPath ~/.ssh/%r@%h:%p
```

**Comment by h2k82u** [comment]
@etz1  and @kevinashworth:
sorry to ask noob questions, but I don't get it working
- the ssh config has to be on the laptop, desktop or both?
- does it work for you using an IP adress?
I sync on local network only and use the local IP of my desktop (192.168.178.7) in the unison .prf file on my laptop. I am able to ssh to my desktop via terminal but the unison gui is still not working ('connecting...')
Thank you for your help!

**Comment by etz1** [comment]
Config should be under .ssh/ directory on laptop and it should work with numerical IP address, I think

**Comment by kevinashworth** [comment]
Yes, my `config` is in `~/.ssh` on my laptop Mac. I first type the ssh command on my laptop Mac, then I open Unison on my laptop and that's how I can sync my laptop with my desktop Mac. On my laptop -- in both the command line and in Unison settings -- I'm using the ssh command that's presented to me under my desktop Mac's System Preferences > Sharing > Remote Login. HTH!

**Comment by h2k82u** [comment]
@etz1 and @kevinashworth: thank you for your fast response, I really appreciate it!
I followed Kevin's instructions but without any success.
Do you use a passphrase for ssh (thats what I do) or an authentication key?

What I don't understand is that Unison was working in the exact setting for more than 5 years until the last Catalina update

**Comment by etz1** [comment]
If you follow these instructions, login to the remote computer, and then open another terminal in the laptop and ssh, does it require a pw? It should not...
It seems there is a problem with unison now, this procedure gets around it by eliminating the need for a pw rather than solving it

**Comment by h2k82u** [comment]
ok, that seems to be the problem.
I ssh to my desktop using a password without any problem ('ssh myacount@10.0.1.7').
After opening a second terminal window I repeat the ssh command but I am asked for the password again.
As suggested I created a 'config' file in ~/.ssh (I only had a file named 'ssh_config' in /etc/ssh/) before. It contains the following lines only:
```
Host *
   ControlMaster auto
   ControlPath ~/.ssh/%r@%h:%p
```
I guess, I have to learn more about ssh usage and pitfalls...

**Comment by h2k82u** [comment]
finally got it working!
The key is to multiplex ssh connections as suggested by @etz1 and @kevinashworth.
I had to modify the ssh config a bit differently and created a subdirectory for the 'ControlPath' as described here in detail (https://www.anchor.com.au/blog/2010/02/ssh-controlmaster-the-good-the-bad-the-ugly/)
After that I was able to establish a second ssh connection without being asked for a password and the unison gui was working as well (but I have to ssh in before).
Thank you for you help!
Maybe someone with programming skills is able to fix unison completely...

**Comment by mkeitsch** [comment]
My solution was to simply add my public ssh key of my "main unison sync mac" (where I start the GUI) to the `~/.ssh/authorized_keys` file of my "slave unison sync mac".

For this solution working on my Macs...
- no password entering needed anymore
- no text sync needed
- no unison install via brew or macports or something else needed
- no `.ssh/config` needed
- no ssh multiplexing needed

If this simple solution does not work for others, maybe a combination of the different approaches may help.

I think unison has some issues with piping/recognizing the password prompt of the ssh connection/command or something like this. But I'm not that familiar with GUI programming and connecting GUIs to cli commands. So it is very speculative.

**Comment by gdt** [comment]
On the assumption that Apple has diverged from OpenSSH, I have labeled this an impact-low enhancement.  We can relabel it an defect if the problem also appears with actual OpenSSH.

**Comment by tleedjarv** [comment]
> If somebody wants to do a little objective C programming, one could try to protect all the calls to raisePasswordWindow with a little loop that throws away empty prompts from ssh.

The looping is already done in `terminal.ml` but it does not work in this case for some reason. To fix this, the exact output from the ssh process is needed. I don't have access to macOS. If someone who has had this issue could run ssh manually and collect the exact output (all the non-visible whitespace included, etc.), that would help a lot.

For time being, I've opened a PR which may solve the issue. The looping in `terminal.ml` only skips empty prompts by checking `\r\n`. I'm not quite sure why that is. I've added a check for `\n` to see if this would make any difference.

A request to everyone who has had or still has this issue, if you could test the fix from #465 (you can download binaries from https://github.com/bcpierce00/unison/actions/runs/486355550 (scroll down and look for "Artifacts")).

**Comment by eli-tziperman** [comment]
I would love to help, as I use unison a few times per day and this is an
issue I have to deal with all the time. please tell me exactly which
command to give in a terminal and I will send the output. cheers, Eli

**Comment by tleedjarv** [comment]
Thank you, Eli.

The exact command unison uses is
``` sh
ssh -l username 192.168.0.0 -e none unison -server
```
(substitute your username and IP/hostname).

But for purposes of this issue the `unison -server` part is not relevant and any invocation format should suffice, for example, I would expect all the following to yield the same result:
``` sh
ssh -l username 192.168.0.0
ssh username@192.168.0.0
ssh 192.168.0.0
```
(substitute your username and IP/hostname).

**Comment by eli-tziperman** [comment]
great, thanks here is what I am getting:
$ ssh -l XX XX -e none unison -server
Password:
Warning: No xauth data; using fake authentication data for X11 forwarding.
X11 forwarding request failed on channel 1
2021-01-15 05:53:55.802 Unison[80634:32806919] Calling nonGuiStartup
Unison 2.51 with OCaml >= 4.01.2

**Comment by tleedjarv** [comment]
Thank you very much. Can you try once more and add `-x` to ssh. For example, `ssh -x -l ...`

Also, if you could try a binary from this page https://github.com/bcpierce00/unison/actions/runs/486355550 (I think you have to be logged in to GitHub to see downloads) and see if it hangs in GUI or not.

**Comment by tleedjarv** [comment]
And one more thing, if you can. In GUI, could you add in your profile preference `sshargs = -x` and see if that makes any difference.

Also, for the record, what is your exact OS version and ssh version? (`ssh -V`)

**Comment by gdt** [comment]
Is ssh on macos really trying to forward X11 when the user didn't ask for it? Does macos come with an X11 server now?

**Comment by eli-tziperman** [comment]
Here it is, hope I got it right:

$ ssh -l XX XX -e none unison -server
Password:
Warning: No xauth data; using fake authentication data for X11 forwarding.
X11 forwarding request failed on channel 1
2021-01-15 09:32:25.639 Unison[81238:32839241] Calling nonGuiStartup
Unison 2.51 with OCaml >= 4.01.2

$ ssh -l XX XX -x -e none unison -server
Password:
2021-01-15 09:32:44.403 Unison[81245:32839335] Calling nonGuiStartup
Unison 2.51 with OCaml >= 4.01.2

$ ssh -V
OpenSSH_8.3p1, OpenSSL 1.1.1h  22 Sep 2020

also, I am using
MacOS Catalina (10.15.7)
and I have XQuartz installed

when I add
sshargs = -x
to the .unison profile file it does not seem to solve the problem.

when I try the latest binary from there I get an error message:

*“Unison.app” is damaged and can’t be opened. You should move it to the
Trash.*


cheers, Eli

**Comment by eli-tziperman** [comment]
regarding the latest binary, I take that back. when I tried again it does
not unzip into an app. should I run it from the terminal?

**Comment by tleedjarv** [comment]
There is a tar.gz inside a zip file. You have to extract both to get the app.
To get past the "app is damaged" error, you have to remove the quarantine xattr (see https://github.com/bcpierce00/unison/wiki/CI-Binary-instructions).

**Comment by tleedjarv** [comment]
Thank you for help with testing, Eli. Let's wait and see if you can try the app from the PR.

**Comment by eli-tziperman** [comment]
got it, thanks.  so when I open the app version,
2.51.70, ocaml 4.11.1
it hangs up and does not request password. if I first ssh to the server
from a terminal so that password is not needed, it works great (then
complains that the server is an older unison version, which I think can be
ignored at this point).  Eli

**Comment by tleedjarv** [comment]
Could you also see in the log (or start from terminal) if there is the error message "Unrecognized message from ssh".

**Comment by eli-tziperman** [comment]
nothing in the unison log file or in output to terminal, but I added an -E
option to the sshargs in the unison profile file and the log file for that
gave this:

OpenSSH_8.3p1, OpenSSL 1.1.1h  22 Sep 2020
debug1: Reading configuration data /Users/XX/.ssh/config
debug1: /Users/XX/.ssh/config line 75: Applying options for XX.edu
debug1: Reading configuration data /usr/local/etc/ssh/ssh_config
debug1: auto-mux: Trying existing master
debug1: Control socket "/Users/XX/.ssh/XX@XX.edu:22" does not exist
debug1: Connecting to XX.edu [140.247.105.27] port 22.
debug1: Connection established.
debug1: identity file /Users/XX/.ssh/id_rsa type 0
debug1: identity file /Users/XX/.ssh/id_rsa-cert type -1
debug1: identity file /Users/XX/.ssh/id_dsa type -1
debug1: identity file /Users/XX/.ssh/id_dsa-cert type -1
debug1: identity file /Users/XX/.ssh/id_ecdsa type -1
debug1: identity file /Users/XX/.ssh/id_ecdsa-cert type -1
debug1: identity file /Users/XX/.ssh/id_ecdsa_sk type -1
debug1: identity file /Users/XX/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /Users/XX/.ssh/id_ed25519 type -1
debug1: identity file /Users/XX/.ssh/id_ed25519-cert type -1
debug1: identity file /Users/XX/.ssh/id_ed25519_sk type -1
debug1: identity file /Users/XX/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /Users/XX/.ssh/id_xmss type -1
debug1: identity file /Users/XX/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.3
debug1: Remote protocol version 2.0, remote software version OpenSSH_7.9
debug1: match: OpenSSH_7.9 pat OpenSSH* compat 0x04000000
debug1: Authenticating to XX.edu:22 as 'eli'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC:
<implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC:
<implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256 XX:XX
debug1: Host 'XX.edu' is known and matches the ECDSA host key.
debug1: Found key in /Users/XX/.ssh/known_hosts:10
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /Users/XX/.ssh/id_rsa RSA SHA256:XX
debug1: Will attempt key: /Users/XX/.ssh/id_dsa
debug1: Will attempt key: /Users/XX/.ssh/id_ecdsa
debug1: Will attempt key: /Users/XX/.ssh/id_ecdsa_sk
debug1: Will attempt key: /Users/XX/.ssh/id_ed25519
debug1: Will attempt key: /Users/XX/.ssh/id_ed25519_sk
debug1: Will attempt key: /Users/XX/.ssh/id_xmss
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info:
server-sig-algs=<ssh-ed25519,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue:
publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering public key: /Users/XX/.ssh/id_rsa RSA SHA256:XX
debug1: Authentications that can continue:
publickey,password,keyboard-interactive
debug1: Trying private key: /Users/XX/.ssh/id_dsa
debug1: Trying private key: /Users/XX/.ssh/id_ecdsa
debug1: Trying private key: /Users/XX/.ssh/id_ecdsa_sk
debug1: Trying private key: /Users/XX/.ssh/id_ed25519
debug1: Trying private key: /Users/XX/.ssh/id_ed25519_sk
debug1: Trying private key: /Users/XX/.ssh/id_xmss
debug1: Next authentication method: keyboard-interactive

**Comment by tleedjarv** [comment]
Based on the information so far, there is nothing clearly at fault.

Other testers are welcome to collect more information.

**Comment by eli-tziperman** [comment]
perhaps this would help:  if I ssh to the server first so that there is no
need to enter password and unison works well, the ssh log file is:


OpenSSH_8.1p1, LibreSSL 2.7.3
debug1: Reading configuration data /Users/XX/.ssh/config
debug1: /Users/XX/.ssh/config line 75: Applying options for XX.edu
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 47: Applying options for *
debug1: auto-mux: Trying existing master


if I do not ssh first, one gets the much longer ssh log file I sent
previously, here it is again:

OpenSSH_8.1p1, LibreSSL 2.7.3
debug1: Reading configuration data /Users/XX/.ssh/config
debug1: /Users/XX/.ssh/config line 75: Applying options for XX.edu
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 47: Applying options for *
debug1: auto-mux: Trying existing master
debug1: Control socket "/Users/XX/.ssh/XX@XX.edu:22" does not exist
debug1: Connecting to XX.edu port 22.
debug1: Connection established.
debug1: identity file /Users/XX/.ssh/id_rsa type 0
debug1: identity file /Users/XX/.ssh/id_rsa-cert type -1
debug1: identity file /Users/XX/.ssh/id_dsa type -1
debug1: identity file /Users/XX/.ssh/id_dsa-cert type -1
debug1: identity file /Users/XX/.ssh/id_ecdsa type -1
debug1: identity file /Users/XX/.ssh/id_ecdsa-cert type -1
debug1: identity file /Users/XX/.ssh/id_ed25519 type -1
debug1: identity file /Users/XX/.ssh/id_ed25519-cert type -1
debug1: identity file /Users/XX/.ssh/id_xmss type -1
debug1: identity file /Users/XX/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.1
debug1: Remote protocol version 2.0, remote software version OpenSSH_7.9
debug1: match: OpenSSH_7.9 pat OpenSSH* compat 0x04000000
debug1: Authenticating to XX.edu:22 as 'XX'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC:
<implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC:
<implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:XX
debug1: Host 'XX.edu' is known and matches the ECDSA host key.
debug1: Found key in /Users/XX/.ssh/known_hosts:19
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /Users/XX/.ssh/id_rsa RSA SHA256:XX
debug1: Will attempt key: /Users/XX/.ssh/id_dsa
debug1: Will attempt key: /Users/XX/.ssh/id_ecdsa
debug1: Will attempt key: /Users/XX/.ssh/id_ed25519
debug1: Will attempt key: /Users/XX/.ssh/id_xmss
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info:
server-sig-algs=<ssh-ed25519,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue:
publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering public key: /Users/XX/.ssh/id_rsa RSA SHA256:XX
debug1: Authentications that can continue:
publickey,password,keyboard-interactive
debug1: Trying private key: /Users/XX/.ssh/id_dsa
debug1: Trying private key: /Users/XX/.ssh/id_ecdsa
debug1: Trying private key: /Users/XX/.ssh/id_ed25519
debug1: Trying private key: /Users/XX/.ssh/id_xmss
debug1: Next authentication method: keyboard-interactive

**Comment by gdt** [comment]
In your long log, pubkey auth fails.   I think you should adjust your ssh setup so that something like "ssh remotehost id" just works.   Perhaps there is a notion of the underlying ssh from unison prompting for and getting a password, but I would not expect that to work reliably.

**Comment by tleedjarv** [comment]
It would be very helpful if someone experiencing this issue could get a stack trace of the hanging process.

**Comment by eli-tziperman** [comment]
happy to try if you could tell me exactly what to do

**Comment by tleedjarv** [comment]
As I don't have access to macOS myself, I can't really tell. I'm sure there are multiple ways, but this seems to work directly in GUI: https://apple.stackexchange.com/questions/272508/how-do-you-get-a-stack-trace-of-a-running-process

**Comment by eli-tziperman** [comment]
done.  may I email this to you directly rather than post? not sure how much
private information is there...

**Comment by tleedjarv** [comment]
Big thank you to @eli-tziperman for help with testing and debugging. I think the bug has now been found and fixed.

I invite everyone who experienced this issue to test the solution available in PR #465 (you can do this by downloading the built app from here https://github.com/bcpierce00/unison/actions/runs/763358278).

There is a tar.gz inside a zip file. You have to extract both to get the app.
To get past the "app is damaged" error, you have to remove the quarantine xattr (see https://github.com/bcpierce00/unison/wiki/CI-Binary-instructions).

**Comment by matatata** [comment]
Hi everybody,
I just want to add that I encountered a problem like this when I switched from 2.15.4 to 2.51.5 (on Big Sur) today. I'm using Kerberos btw:
`sshargs = -K`

I found this discussion and changing
`root = ssh://hostname`
to
`root = ssh://usernamr@hostname`
luckily solved the issue for me.

**Comment by tleedjarv** [comment]
Can you please post the exact error message you are getting?

**Comment by thetrial** [comment]
`Unrecognized message from ssh: '2022-01-18 18:45:27.010 Unison[3784:121413] Calling nonGuiStartup
'`

**Comment by thetrial** [comment]
And above: Connection Error.

This is the exact message in a GUI window, confirmable with OK.

**Comment by tleedjarv** [comment]
Thanks. Yes, this is clearly a bug.

**Comment by tleedjarv** [comment]
@thetrial are you able to test with 2.51.4 and confirm if the same issue is present with that version as well?

**Comment by tleedjarv** [comment]
@thetrial I don't know if this is the cause of the issue you're seeing but I have spotted one possible bug for which I have a tentative fix. Do you think you can grab a build from https://github.com/tleedjarv/unison/actions/runs/1714585503 and test it out? See instructions for how to grab the build a few comments above.

**Comment by thetrial** [comment]
I grabbed that and put it on one Mac, with xattr -cr procedure and exported binary to /usr. I can test the second tomorrow. At least it syncs with the Rasperry Pi. I was just a little irritated why I use exactly Unison 2.51.5 (ocaml 4.11.2) on the Mac(s) – due to the fact that the linux version is 2.51.3 (ocaml 4.11.1). Then I saw that there was no (is no?) GUI app for mac in this version. So the next near seemed to me 2.51.5 (ocaml 4.11.2). Is that really compatible?

But for tomorrow I’ll test this same named new build, Mac agains Mac. Shall I then test 2.51.4, too? Or not? Or is there a 2.51.3 as app?

**Comment by thetrial** [comment]
Seems not to work either. My first attempt went through. The second came once with the error, then it worked. My third profile did not go through.

`Unrecognized message from ssh: '2022-01-19 08:14:00.903 Unison[4879:164528] Calling nonGuiStartup
'`

**Comment by thetrial** [comment]
Hereafter I tried 2.51.4 (ocaml 4.11.2) … and for now it seems, that this works! There are no 2.51.3 Mac apps, are there? Is there an overview which ocaml versions match and witch don’t?

**Comment by tleedjarv** [comment]
Don't worry about the versions, they are expected to be fully compatible. Even OCaml versions are mostly compatible, even different minor versions and definitely different patch versions.

Ok, so far we have gotten some important clues here. According to your testing, the issue appeared in the 2.51.5 release. Let's try to pin it down a little further. I started another test build at https://github.com/tleedjarv/unison/actions/runs/1716928058 When that finishes, it would be great if you could test it.

**Comment by thetrial** [comment]
Of course I would.

**Comment by thetrial** [comment]
2bd6bfca gives the error again:

`Unrecognized message from ssh: '2022-01-19 17:17:35.965 Unison[5472:190103] Calling nonGuiStartup
'`

**Comment by tleedjarv** [comment]
I've started two new debug builds. Could you try each of them and report the results? (I don't have access to any macOS, that's why we're going about it like this...)
https://github.com/tleedjarv/unison/actions/runs/1719084788
https://github.com/tleedjarv/unison/actions/runs/1719105153

**Comment by thetrial** [comment]
Test at dusk.
406db4cc has the error.
`Unrecognized message from ssh: '2022-01-20 06:56:40.527 Unison[32229:316247] Calling nonGuiStartup
'`
f3d5eb4b … first attempts seem to run through. No error yet.
This was just a quick test without installing to /usr and only on one Mac, to sync to the second. Should more thorough tests follow?

**Comment by thetrial** [comment]
By the way, I should mention a thing from the past. My machines connect with exchanged ssh keys. After I recreated the keys long time ago, a small input window occurred and asked for the password. I don’t know if for the key or because something with the key went wrong. I guess (but don’t know) it was deposit in the system keychain after that. Before I updated Unison, everything worked without query. I don’t know if this is relevant, but it should be mentioned for possible future scenarios.

**Comment by tleedjarv** [comment]
Thank you for testing. I've now confirmed where the issue is but the cause is still a bit unclear.

The cause seems to be `NSLog` outputing to stderr and that output is read back in as if it was output by the ssh process. https://developer.apple.com/documentation/foundation/1395074-nslogv specifies that `NSLog` does output to stderr under certain conditions. What is unclear to me is why `NSLog` output from the **parent Unison process** is written to stderr/tty of the **ssh child process**. If anyone has a clue, please do share it here.

For now, I have started another test build for you to try out. It works around (hopefully) the issue by removing the `NSLog` that is producing this `Calling nonGuiStartup` message.

https://github.com/tleedjarv/unison/actions/runs/1723443199 Could you please test this one and see if it makes a difference?

> Before I updated Unison, everything worked without query.

I understand this as updating from 2.51.4 to 2.51.5, is that correct? Did the query appear only once or does it appear now on every run?

**Comment by gdt** [comment]
I have a fuzzy memory of some other problem with ssh and file descriptors ending up being shared somehow when you wouldn't expect them to be.   I am not sure it was in unison, or on macOS, and reading POSIX there was something about cloning a fd on fork vs cloning the underlying "file description object" or some words like that -- it was confusing.

**Comment by thetrial** [comment]
> >    Before I updated Unison, everything worked without query.
>
> I understand this as updating from 2.51.4 to 2.51.5, is that correct? Did the query appear only once or does it appear now on every run?

No, I meant updating Unison from 2.48.15 (which I used a very long time) to 2.51.*.

I will check your new version.

**Comment by thetrial** [comment]
Sorry to say …
a265ec91, again has the error:
`Unrecognized message from ssh: '2022-01-20 17:35:18.956 Unison[32668:336170] Calling nonGuiStartup
'`

**Comment by tleedjarv** [comment]
Well this is interesting... As you can see in the commit, the message `Calling nonGuiStartup` is clearly disabled; this binary should never output this message. So where is it coming from? I guess you have multiple binaries around?

**Comment by thetrial** [comment]
I have the working 2.51.4 and its cli binary under /usr. I did not delete them before. I started the test version from the download directory directly. I did not export its cli binary to /usr. But those don’t bring up this message. So … it’s strange?!

**Comment by thetrial** [comment]
So what shall we do now?

**Comment by tleedjarv** [comment]
Can you start the GUI app in the terminal? (I don't know how it works in macOS) Can you try redirecting it's stderr to a file? By Apple's documentation, this should redirect `NSLog`'s output also to that file. Let's see if it makes any difference.

**Comment by tleedjarv** [comment]
Another thing to try is to temporarily move away all other binaries, make sure that the tested binary really is the one with commit a265ec9, and then try again.

**Comment by thetrial** [comment]
Before doing that (I don’t know how yet) I deleted my working Unison, replaced it, exported the cli binary to /usr/local/bin. Still same error with a265ec91.
Check:
```
~ > unison -version
unison version 2.51.5 (ocaml 4.11.2)
```
Starting from terminal:
```
~ > unison -ui graphic
2022-01-20 18:06:51.431 Unison[22449:812176] Unrecognized message from ssh: '2022-01-20 18:06:51.436 Unison[32933:347498] Calling nonGuiStartup
'
```
How would you redirect the stderr to a file with bsd, unix oder linux? I’d try that way here.

**Comment by thetrial** [comment]
Oh … I found out this with locate unison:

```
/usr/bin/unison
/usr/local/bin/unison
```
Hm, which one is right, which one is wrong?!

**Comment by thetrial** [comment]
No, no, forget ist, I guess the locate database is old. There is only /usr/local/bin/unison!

**Comment by thetrial** [comment]
I redirected … but it’s the same output as in the terminal …

I tried three profiles to the second Mac. The first did not give an error (!), the second and third gave:

```
2022-01-20 18:17:18.816 Unison[26003:825265] Unrecognized message from ssh: '2022-01-20 18:17:18.825 Unison[32990:349954] Calling nonGuiStartup
'
2022-01-20 18:17:28.654 Unison[26003:825265] Unrecognized message from ssh: '2022-01-20 18:17:28.663 Unison[32994:349984] Calling nonGuiStartup
'

```

**Comment by thetrial** [comment]
Now I’m back again to 2.51.4 – and everything works.

**Comment by tleedjarv** [comment]
I have to say, I'm at a loss here.

The `Calling nonGuiStartup` message is output by the startup code here (the code is very short, you can just read through it):
https://github.com/bcpierce00/unison/blob/3ac6a7f44e678148c47aed3a2a1a06d2cf30330f/src/uimac/main.m
more specifically: https://github.com/bcpierce00/unison/blob/3ac6a7f44e678148c47aed3a2a1a06d2cf30330f/src/uimac/main.m#L38

The same code will then start the real GUI, here: https://github.com/bcpierce00/unison/blob/3ac6a7f44e678148c47aed3a2a1a06d2cf30330f/src/uimac/main.m#L52

which will then in turn request starting the ssh process and then interact with it. The code for that is here: https://github.com/bcpierce00/unison/blob/3ac6a7f44e678148c47aed3a2a1a06d2cf30330f/src/uimac/MyController.m#L407-L519

All of this is done in a proper Mac "app" (I don't know what's the proper way to call it), the one in Objective C. The `unison` that you invoke on the command line is a "regular" Unix executable, the plain C code for which is here (the code is very short, you can just read through it): https://github.com/bcpierce00/unison/blob/3ac6a7f44e678148c47aed3a2a1a06d2cf30330f/src/uimac/cltool.c
All it does is look for the "app" and start it.

I suspect that somehow an incorrect copy of the "app" is found... ?? Is there a way to start the "app" directly, skipping the `unison` binary completely? There must be, because the `unison` binary from `cltool.c` is provided for convenience only, not as a required component.

**Comment by thetrial** [comment]
What do you mean with incorrect copy of the "app"? Which app? The "thing" that is packed and has an icon to click on it? Why should this be incorrect? And why is version 2.51.4 starting without any problems?
I don’t start unison via terminal the normal way, I start the "app". I thought this graphical app starts the unison binary "inside" itself. And the unix binary under /usr/local/bin (/local!) is just used by the incoming unison connection of another machine.

Well, I don’t understand why it works with one and not with the other. What is the diff?

**Comment by tleedjarv** [comment]
Yes, that's what I mean by "app", that "thing" :) I don't know the correct terminology. I have zero experience with macOS. @gdt maybe you have some guidance here?

When downloading the binary that includes commit a265ec9, a simple search for string `Calling nonGuiStartup` returns nothing in that binary (the one named `Unison`, not `unison`). So that binary does not produce that output. A search in other binaries (without this commit), easily shows that the string `Calling nonGuiStartup` is present.

Another thing that I didn't notice before because the error message from the GUI contained less information, in the messages like this:
```
2022-01-20 18:17:18.816 Unison[26003:825265] Unrecognized message from ssh: '2022-01-20 18:17:18.825 Unison[32990:349954] Calling nonGuiStartup
'
```

the process IDs `26003:825265` and `32990:349954` are different. Also, what I didn't notice before, the timestamp of the embedded message is _later_ than that of the outer error message! This is significant because the embedded message causes the outer error and should never have a later timestamp (even with potential buffering going on). So I'm wondering if these are really separate processes on the same machine or could the embedded message come from the other machine (???).

When you receive this error in the GUI, before closing the message window, could you take a snapshot of the process list to see if these are processes on the same machine and what commands are behind them. Sorry, I can't tell you how to view a process list on macOS. Also, you could snapshot the process list on the other machine to see if the other process ID is present there.

**Comment by gdt** [comment]
The only really reliable way to be sure one is testing the right code is to remove every single vestige of unison except for what it is being tested.   Also, I think it's best to test with command line unison if that can exhibit the bug.

Also, it makes sense to do the same purge and install only the testing version on the remote side too.

**Comment by tleedjarv** [comment]
I finally figured out what's going on. The "unrecognized message" _is_ in fact coming from the remote machine's stderr. I'll see if I can find a fix.

**Comment by tleedjarv** [comment]
I've opened a PR with a proposed fix. @thetrial could you please test a build from https://github.com/bcpierce00/unison/actions/runs/1728155694
I don't have any access to macOS, so I am doing the fix completely in the blind. Your testing is very important to get the fix merged.

If you can, please test as follows:
1) local machine with the fixed build --> remote machine with 2.51.5
2) local machine with the fixed build --> remote machine with 2.51.4
3) local machine with the fixed build --> remote machine with the fixed build

Not for others willing to test -- this issue and the fix applies _only_ to the uimac GUI, which is the native Mac GUI. It does not affect the GTK GUI.

**Comment by thetrial** [comment]
I’m going to check these setups in the next two days. Hold the line.

**Comment by tleedjarv** [comment]
Thank you for your testing efforts. Based on the results, I will mark the PR as ready to merge.

The Linux crash most likely is unrelated, but it is good to know that real error messages from ssh are properly presented to the user.

**Comment by reagle** [comment]
This issue is closed, but I'm experiencing this issue with 2.51.5 (from homebrew). Is there a fix pending in the next version? Thanks to this thread, if I ssh to the other server in a separate terminal first, it works, but otherwise I can only use the text UI.

```
╰─➤  unison -ui graphic
2022-01-28 15:36:20.108 Unison[49055:369136] Calling nonGuiStartup
╰─➤  unison -version
2022-01-28 15:36:30.367 Unison[49067:369291] Calling nonGuiStartup
unison version 2.51.5 (ocaml 4.12.0)
```

**Comment by tleedjarv** [comment]
Yes, the fix is not released yet. If you want, you can try out the fix by downloading a pre-built binary at https://github.com/bcpierce00/unison/actions/runs/1733125753 and you have to follow guidelines at https://github.com/bcpierce00/unison/wiki/CI-Binary-instructions
If this works for you then you can keep using this unreleased version. If this doesn't fix your issue then please report more details about your exact issue.

**Comment by reagle** [comment]
Which one of those many files would I download? I'm running macOS 12.2 21D49 arm64 -- and my ocaml version is above.

**Comment by tleedjarv** [comment]
The macOS GUI: https://github.com/bcpierce00/unison/suites/5026190466/artifacts/148102599
The "normal" Unix binary and GTK GUI (just FYI, you don't need this, the one above should be all you need): https://github.com/bcpierce00/unison/suites/5026190466/artifacts/148102598

**Comment by reagle** [comment]
I recently installed a new system and am having this problem again. Is a new release close?

**Comment by gdt** [comment]
Homebrew not having the latest releaes is not a unison bug.
